src/logic/mul.rs

```diff
@@ -0,0 +1,395 @@
+                            return Ok(())
+                        }
+                    },
+                    (true, false) => {
+                        let mult = lhs[0];
+                        //first digit done and carry
+                        let temp = mult.carrying_mul(rhs[0]);
+                        lhs[0] = temp.0;
+                        let mut mul_carry = temp.1;
+                        //middle of row
+                        for rhs_i in 1..rhs_sig_nonzero {
+                            let temp = mult.carrying_mul_add(rhs[rhs_i], mul_carry);
+                            lhs[rhs_i] = temp.0;
+                            mul_carry = temp.1;
+                        }
+                        //final digit
+                        if rhs_sig_nonzero == lhs.len() - 1 {
+                            lhs[rhs_sig_nonzero] = mult
+                                .wrapping_mul_add(rhs[rhs_sig_nonzero], mul_carry);
+                        } else {
+                            let temp = mult.carrying_mul_add(rhs[rhs_sig_nonzero], mul_carry);
+                            lhs[rhs_sig_nonzero] = temp.0;
+                            lhs[rhs_sig_nonzero + 1] = temp.1;
+                        }
+                    },
+                    (false, true) => {
+                        //first digit done and carry
+                        let temp = rhs[0].carrying_mul(lhs[0]);
+                        lhs[0] = temp.0;
+                        let mut mul_carry = temp.1;
+                        //middle of row
+                        for lhs_i in 1..lhs_sig_nonzero {
+                            let temp = rhs[0].carrying_mul_add(lhs[lhs_i], mul_carry);
+                            lhs[lhs_i] = temp.0;
+                            mul_carry = temp.1;
+                        }
+                        //final digit
+                        if lhs_sig_nonzero == lhs.len() - 1 {
+                            lhs[lhs_sig_nonzero] = rhs[0]
+                                .wrapping_mul_add(lhs[lhs_sig_nonzero], mul_carry);
+                        } else {
+                            let temp = rhs[0].carrying_mul_add(lhs[lhs_sig_nonzero], mul_carry);
+                            lhs[lhs_sig_nonzero] = temp.0;
+                            lhs[lhs_sig_nonzero + 1] = temp.1;
+                        }
+                    },
+                    (true, true) => {
+                        let temp0 = lhs[0].carrying_mul(rhs[0]);
+                        lhs[0] = temp0.0;
+                        lhs[1] = temp0.1;
+                    }
+                }
+            }
+        }
+        self.clear_unused_bits();
+        Ok(())
+    }
+
+    /// Multiplies `rhs` with `self` and returns the result.
+    pub fn into_wrapping_mul(self, rhs: &ApInt) -> Result<ApInt> {
+        try_forward_bin_mut_impl(self, rhs, ApInt::wrapping_mul_assign)
+    }
+}
+
+#[cfg(test)]
+#[rustfmt::skip]
```

Comment:
Always add a rational if you do things like skipping rust format, especially for such huge blocks.

Reply:
... or dont do things like these ;)